### ISSUE TITLE
Difference between NLVR2 Dev performance

### ISSUE THREAD

**Issue by sanjayss34** [airsplay/lxmert#5]
When you ran the NLVR2 fine-tuning and reported your result in this issue https://github.com/airsplay/lxmert/issues/1#issuecomment-523983982, the result (dev performance) was 74.39. Similarly, when I ran fine-tuning, I got about 74.33 for dev accuracy. How come the dev accuracy reported in the paper, in the table in the README, and in the NLVR2 leaderboard (http://lil.nlp.cornell.edu/nlvr/) is 74.9? Thanks for your help.

**Comment by airsplay** [comment]
Hi, since some CuDNN operations are not deterministic (see [here](https://pytorch.org/docs/stable/notes/randomness.html) for a reference), the results for multiple runs are different even I already fixed all the seeds [here](https://github.com/airsplay/lxmert/blob/8aade4657c0671fec23f6fb26fd17c0541d053f8/src/param.py#L97) (random seed, np.random seed, torch seed which would set both torch CPU and torch GPU seeds). 

The leaderboard snapshot, your run, and my previous run in [comment](https://github.com/airsplay/lxmert/issues/1#issuecomment-523983982) would be different from each other and the range of difference also depends on the size of validation/test set.

As a result, on VQA(5K images) /GQA(300 images) /NLVR2 (2K images) validation set, the results vary a little bit.  But the results on VQA/GQA test sets (with around 10 to 50K images) are much more stable. We can not actually verify it because we only have 5 submission entries. But our results for two different runs only vary less than 0.1 on VQA/GQA test set.

Hope this helps!

**Comment by airsplay** [comment]
By the way, we actually got around 59~60 result on NLVR2 early this year, but we do not think

**Comment by sanjayss34** [comment]
Thanks for your response. For the results when you got 74.9, did you use the random seed that you have written in param.py, or did you run several times (without fixing the seed) and choose the run in which you got the best result on the validation set? I'm asking only about NLVR2.

**Comment by airsplay** [comment]
The two results are run with the same code and the same seeds. The only difference is that they are two different runs (type the same commands and press `enter` in two different days).

The reason for these different results is that the GPU computation is non-deterministic. Moreover, it is due to that the [associativity](https://en.wikipedia.org/wiki/Associative_property) is not held in floating-number numerical computation, which means `(a+b) + c` may not equal to `a + (b + c)`. As a result, the result may differ if the order of the computation changes. 

We did not tune the results based on the seed. This kind of behavior is not acceptable to me at all.

**Comment by sanjayss34** [comment]
Thanks for your detailed response. That makes sense.
